cartographer/mapping/internal/3d/pose_graph_3d_test.cc

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/mapping/internal/3d/pose_graph_3d.h"
+
+#include "cartographer/mapping/internal/test_helpers.h"
+#include "cartographer/mapping/proto/serialization.pb.h"
+#include "cartographer/transform/rigid_transform.h"
+#include "cartographer/transform/rigid_transform_test_helpers.h"
+#include "cartographer/transform/transform.h"
+#include "gmock/gmock.h"
+#include "google/protobuf/util/message_differencer.h"
+
+namespace cartographer {
+namespace mapping {
+namespace {
+
+class PoseGraph3DForTesting : public PoseGraph3D {
+ public:
+  PoseGraph3DForTesting(const proto::PoseGraphOptions& options,
+                        common::ThreadPool* thread_pool)
+      : PoseGraph3D(options, thread_pool) {}
+
+  void WaitForAllComputations() { PoseGraph3D::WaitForAllComputations(); }
+};
+
+class TestEnvironment {
+ public:
+  TestEnvironment()
+      : thread_pool_(common::make_unique<common::ThreadPool>(1)) {}
+
+  TestEnvironment& SetDefault() {
+    const std::string kPoseGraphLua = R"text(
+      include "pose_graph.lua"
+      return POSE_GRAPH)text";
+    auto pose_graph_parameters = test::ResolveLuaParameters(kPoseGraphLua);
+    pose_graph_options_ = CreatePoseGraphOptions(pose_graph_parameters.get());
+    return *this;
+  }
+
+  std::unique_ptr<PoseGraph3DForTesting> Build() {
+    return common::make_unique<PoseGraph3DForTesting>(pose_graph_options_,
+                                                      thread_pool_.get());
+  }
```

Comment:
Can `TestEnvironment` be replaced with a test fixture `PoseGraph3DTest` that contains `std::unique_ptr<PoseGraph3D>` and initializes itself in constructor or `SetUp` method?

Reply:
Done